components/nexus-repository/src/main/java/org/sonatype/nexus/repository/httpclient/FilteredHttpClient.java

```diff
@@ -0,0 +1,354 @@
+ *
+ * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
+ * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
+ *
+ * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
+ * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
+ * Eclipse Foundation. All other trademarks are the property of their respective owners.
+ */
+package org.sonatype.nexus.repository.httpclient;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.URI;
+
+import javax.net.ssl.SSLPeerUnverifiedException;
+
+import org.sonatype.nexus.common.sequence.FibonacciNumberSequence;
+import org.sonatype.nexus.common.sequence.NumberSequence;
+import org.sonatype.sisu.goodies.common.ComponentSupport;
+import org.sonatype.sisu.goodies.common.Time;
+
+import org.apache.http.HttpHost;
+import org.apache.http.HttpRequest;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.ClientProtocolException;
+import org.apache.http.client.HttpClient;
+import org.apache.http.client.ResponseHandler;
+import org.apache.http.client.methods.HttpHead;
+import org.apache.http.client.methods.HttpUriRequest;
+import org.apache.http.client.utils.URIUtils;
+import org.apache.http.conn.ClientConnectionManager;
+import org.apache.http.conn.ConnectionPoolTimeoutException;
+import org.apache.http.params.HttpParams;
+import org.apache.http.protocol.HttpContext;
+import org.joda.time.DateTime;
+import org.joda.time.Duration;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * @since 3.0
+ */
+public class FilteredHttpClient
+    extends ComponentSupport
+    implements HttpClient, Closeable
+{
+
+  private final HttpClient delegate;
+
+  private final boolean blocked;
+
+  private HttpHost mainTarget;
+
+  private DateTime blockedUntil;
+
+  private Thread checkThread;
+
+  private final boolean autoBlock;
+
+  private final NumberSequence autoBlockSequence;
+
+  private RemoteConnectionStatus status;
+
+  public FilteredHttpClient(final HttpClient delegate,
+                            final HttpClientConfig config)
+  {
+    this.delegate = checkNotNull(delegate);
+    blocked = checkNotNull(config).getConnectionConfig().isBlocked();
+    autoBlock = config.getConnectionConfig().shouldAutoBlock();
+    status = new RemoteConnectionStatus(blocked ? "Remote Manually Blocked" : "Unknown");
+    // TODO shall we use config.getConnectionConfig().getTimeout() * 2 as in NX2?
+    autoBlockSequence = new FibonacciNumberSequence(Time.seconds(40).toMillis());
+  }
+
+  private <T> T filter(final HttpHost target, final Filterable<T> filterable) throws IOException {
+    // main target is the first accessed target
+    if (mainTarget == null) {
```

Comment:
I did not want to sync on this as in my view different threads for same repo they all start with same target so regardless the task they should all end up with same main target